scalameta/quasiquotes/shared/src/main/scala-3/scala/meta/internal/quasiquotes/ReificationMacros.scala

```diff
@@ -0,0 +1,688 @@
+            maybeHole.getOrElse {
+              val message = "this code path should've been unreachable"
+              val relevantValues = s"""
+                                      |quasi = $quasi
+                                      |quasi.pos.absolutize = ${quasi.pos.absolutize}
+                                      |holes = $holes
+                                      |""".stripMargin
+              throw new UnreachableError(message + relevantValues)
+            }
+      }
+    }
+
+    mode match {
+      case Mode.Term(_, _) =>
+        val internalResult = Internal.liftTree(meta).asInstanceOf[Term].asExpr
+        if (sys.props("quasiquote.debug") != null) println(internalResult)
+        val resType = qType match
+          case QuasiquoteType.Stat => TypeRepr.of[scala.meta.Tree]
+          case QuasiquoteType.TermParam => TypeRepr.of[scala.meta.Term.Param]
+          case QuasiquoteType.Type => TypeRepr.of[scala.meta.Type]
+          case QuasiquoteType.TypeParam => TypeRepr.of[scala.meta.Type.Param]
+          case QuasiquoteType.CaseOrPattern => TypeRepr.of[scala.meta.Tree]
+          case QuasiquoteType.Init => TypeRepr.of[scala.meta.Init]
+          case QuasiquoteType.Self => TypeRepr.of[scala.meta.Self]
+          case QuasiquoteType.Template => TypeRepr.of[scala.meta.Template]
+          case QuasiquoteType.Mod => TypeRepr.of[scala.meta.Mod]
+          case QuasiquoteType.Enumerator => TypeRepr.of[scala.meta.Enumerator]
+          case QuasiquoteType.Importer => TypeRepr.of[scala.meta.Importer]
+          case QuasiquoteType.Importee => TypeRepr.of[scala.meta.Importee]
+          case QuasiquoteType.Source => TypeRepr.of[scala.meta.Source]
+        resType.asType match
+          case '[t] => Block(
+              List(valDef),
+              '{ scala.meta.internal.quasiquotes.Unlift[t]($internalResult) }.asTerm
+            ).asExprOf[t]
+
+      case Mode.Pattern(_, holes, unapplySelector) =>
+        import Internal.internalQuotes.reflect._
+
+        val pattern = Internal.liftTree(meta)
+
+        val args = holes.map(hole => hole.reifier.get.asInstanceOf[Term])
+        val argsContents = args.flatMap {
+          _.tpe match
+            case AppliedType(tpe, content) if tpe =:= TypeRepr.of[Option] => content
+        }
+        val lst = Select.overloaded('{ List }.asTerm, "apply", List(TypeRepr.of[Option[Any]]), args)
+          .asExprOf[List[Option[Any]]]
+
+        val n = args.length
+        val (returned, isBooleanExtractor) =
+          if (n == 0)
+            // boolean extractor
+            ('{ true }, true)
+          else if (n == 1)
+            // returning Tuple1 requires us to manually unapply later, so we return the raw value instead
+            ('{ ${ args.head.asExprOf[Option[Any]] }.get }, false)
+          else (
+            Select.overloaded(
+              Ref(Symbol.classSymbol("scala.Tuple" + n).companionModule),
```

Comment:
Maybe use `TupleClass`?
```suggestion
              Ref(defn.TupleClass(n).companionModule),
```